flash/core/model.py

```diff
@@ -188,3 +210,111 @@ def on_save_checkpoint(self, checkpoint: Dict[str, Any]) -> None:
 
     def configure_finetune_callback(self):
         return []
+
+    ### THE FOLLOWING IS A POC FOR DISTRIBUTED PREDICTION
+    def on_predict_start(self):
+        # TODO: Add hook to lightning Trainer
+        if self.data_pipeline is not None:
+            self.data_pipeline._attach_to_model(self)
+
+        if self.postprocessing_pipeline is not None:
+            self.postprocessing_pipeline._attach_to_model(self)
+
+    def predict_step(self, batch, batch_idx):
+        # TODO: Move lightning predict loop from predict to predict_step
+        if isinstance(batch, (tuple, list)) and len(batch) == 2:
+            x, y = batch
+        else:
+            x, y = batch, None
+
+        return self(x)
+
+    def new_predict(
+        self,
+        x: Any,
+        skip_collate: Optional[bool] = None,
+        data_pipeline: Optional[DataPipeline] = None,
+        postprocessing_pipeline: Optional[PostProcessingPipeline] = None,
+        data_loader_kwargs: Optional[dict] = None,
+        **trainer_kwargs
+    ):
+        if data_pipeline is not None:
+            self.data_pipeline = data_pipeline
+        if postprocessing_pipeline is not None:
+            self.postprocessing_pipeline = postprocessing_pipeline
+
+        trainer = self._create_trainer('predict', **trainer_kwargs)
+
+        if data_loader_kwargs is None:
+            data_loader_kwargs = {}
+
+        if 'num_workers' not in data_loader_kwargs:
+            # leave one for main process
+            data_loader_kwargs['num_workers'] = os.cpu_count() - 1
+
+        auto_collate = None
+        if 'collate_fn' not in data_loader_kwargs:
+            auto_collate = not skip_collate
+
+        dl = self.data_pipeline._generate_loader(x, auto_collate=auto_collate, **data_loader_kwargs)
+
+        return trainer.predict(self, dl)
+
+    def _create_trainer(self, stage: str, **trainer_kwargs):
+        # TODO: Also use these for trainer creation in training?
+        # TODO: Have default trainer kwargs per task?
+        _trainer_kwargs = {}
+        # TODO: Adjust this to trainer running stage from pl
```

Comment:
any thoughts on that @tchaton  @aribornstein ?

Reply:
I like it. We had similar function in previous iteration of predict.

Reply:
I k is we had something like that for training in the beginning. The only downside I See, is That it hides away the Lightning Trainer

Reply:
Maybe we could provide an optional argument for the user to provide trainer in case they don't want to use the default trainer?

Reply:
you mean another trainer class?
Actually, the trainer class is something I'd hardcode here tbh.
This is one of the very fundamental lightning aspects and I feel if a user wants to change it, he either should look into customization with callbacks/plugins or subclass the task to overwrite it here directly.